src/Package/Impl/DataInspect/REnvironmentProvider.cs

```diff
@@ -3,66 +3,96 @@
 
 using System;
 using System.Collections.Generic;
-using System.Diagnostics;
+using System.ComponentModel;
+using System.Linq;
 using System.Threading.Tasks;
 using Microsoft.Common.Core;
 using Microsoft.R.Host.Client;
+using Microsoft.R.StackTracing;
 using Microsoft.VisualStudio.R.Package.Shell;
-using Newtonsoft.Json.Linq;
 
 namespace Microsoft.VisualStudio.R.Package.DataInspect {
-    internal class REnvironmentProvider : IREnvironmentProvider, IDisposable {
+    internal class REnvironmentProvider : IREnvironmentProvider {
+        private static readonly string[] _hiddenEnvironments = { "Autoloads", "rtvs::graphics::ide", ".rtvs" };
+        private static readonly IReadOnlyList<REnvironment> _errorEnvironments = new[] { REnvironment.Error };
+
         private IRSession _rSession;
+        private IREnvironment _selectedEnvironment;
+        private IReadOnlyList<IREnvironment> _environments;
+
+        public event PropertyChangedEventHandler PropertyChanged;
 
         public REnvironmentProvider(IRSession session) {
             _rSession = session;
             _rSession.Mutated += RSession_Mutated;
-
-            GetEnvironmentAsync().DoNotWait();
         }
 
-        private void RSession_Mutated(object sender, EventArgs e) {
-            if (EnvironmentChanged != null) {
-                GetEnvironmentAsync().DoNotWait();
+        public void Dispose() {
+            if (_rSession != null) {
+                _rSession.Mutated -= RSession_Mutated;
+                _rSession = null;
             }
         }
 
-        private async Task GetEnvironmentAsync() {
-            await TaskUtilities.SwitchToBackgroundThread();
+        public IReadOnlyList<IREnvironment> Environments {
```

Comment:
Does it make a difference, given than the entire collection is refreshed completely every time?